data/src/main/java/com/eleks/data/firebase/source/impl/FirebaseDataSourceImpl.kt

```diff
@@ -1,9 +1,177 @@
+        )
+
+    override val groupsFlow = _groupsFlow
+
+    override val userGroupsFlow = _userGroupsFlow
+
+    override val selectedGroupsFlow = _selectedGroupsFlow
+
+    init {
+        launch { subscribeGroups() }
+        launch { subscribeUserGroups() }
+        launch { subscribeSelectedGroups() }
+    }
+
+    private suspend fun subscribeGroups() {
+        var subscription: ListenerRegistration? = null
+        _groupsFlow.subscriptionCount.collect {
+            if (it > 0) {
+                subscription = dbInstance.collection("groups")
+                    .addSnapshotListener(MetadataChanges.INCLUDE) { value, error ->
+                        if (error != null) {
+                            _groupsFlow.tryEmit(ResultDataModel.error(error.message ?: ""))
+                            return@addSnapshotListener
+                        }
+                        if (value?.isEmpty == false) {
+                            val groups = mutableListOf<GroupDataModel>()
+                            for (doc in value) {
+                                groups.add(doc.toObject())
+                            }
+                            groupsFlow.tryEmit(ResultDataModel.success(groups))
+                        }
+                    }
+            } else {
+                subscription?.remove()
+            }
+        }
+    }
+
+    private suspend fun subscribeUserGroups() {
+        var subscription: ListenerRegistration? = null
+        _userGroupsFlow.subscriptionCount.collect {
+            if (authInstance.currentUser == null) return@collect
+            if (it > 0) {
+                subscription = dbInstance.collection("personal")
+                    .document(authInstance.currentUser?.uid ?: "").collection("groups")
+                    .addSnapshotListener { value, error ->
+                        if (error != null) {
+                            _userGroupsFlow.tryEmit(ResultDataModel.error(error.message ?: ""))
+                            return@addSnapshotListener
+                        }
+
+                        if (value?.isEmpty == false) {
+                            val groups = mutableListOf<GroupDataModel>()
+                            for (doc in value) {
+                                groups.add(doc.toObject())
+                            }
+                            _userGroupsFlow.tryEmit(ResultDataModel.success(groups))
+                        }
+                    }
+            } else {
+                subscription?.remove()
+            }
+        }
+    }
+
+    private suspend fun subscribeSelectedGroups() {
+        var subscription: ListenerRegistration? = null
+        _selectedGroupsFlow.subscriptionCount.collect {
+            if (authInstance.currentUser == null) return@collect
+            if (it > 0) {
+                subscription =
+                    dbInstance.collection("personal")
+                        .document(authInstance.currentUser?.uid ?: "")
```

Comment:
small change, but I still suggest to have somethings like this:
`            
val userId = authInstance.currentUser?.uid ?: return@collect
..
 .document(userId).collection("groups")
..
`